File: android/src/main/kotlin/com/valentingrigorean/arcgis_maps_flutter/mapping/symbology/GraphicsController.kt
package com.valentingrigorean.arcgis_maps_flutter.mapping.symbology

import android.content.Context
import com.arcgismaps.mapping.view.Graphic
import com.arcgismaps.mapping.view.GraphicsOverlay
import com.valentingrigorean.arcgis_maps_flutter.convert.map.toGraphicIdValue
import com.valentingrigorean.arcgis_maps_flutter.map.MapTouchGraphicDelegate
import io.flutter.plugin.common.MethodChannel

class GraphicsController (
    private val context: Context,
    private val methodChannel: MethodChannel,
    private val graphicsOverlay: GraphicsOverlay
) : BaseSymbolController(), MapTouchGraphicDelegate {

    private val graphicIdToController: MutableMap<String?, GraphicController> = HashMap()
    private var selectedGraphic: GraphicController? = null

    override fun canConsumeTaps(): Boolean {
        for (controller in graphicIdToController.values) {
            if (controller.consumeTapEvents) {
                return true
            }
        }
        return false
    }

    override fun didHandleGraphic(graphic: Graphic): Boolean {
        val graphicId = graphic.attributes["graphicId"]?.toString() ?: return false

        val graphicController = graphicIdToController[graphicId]
        if (graphicController == null || !graphicController.consumeTapEvents) {
            return false
        }
        if (selectedGraphic != null) selectedGraphic!!.isSelected = false
        graphicController.isSelected = true
        selectedGraphic = graphicController
        invalidateVisibilityFilterController(selectedGraphic!!)
        methodChannel.invokeMethod("graphic#onTap", graphicId.toGraphicIdValue())
        return true
    }

    fun addGraphics(graphicsToAdd: List<Any>?) {
        if (graphicsToAdd == null) {
            return
        }
        val graphicsAdded = ArrayList<Graphic>(graphicsToAdd.size)
        for (graphic in graphicsToAdd) {
            val data = graphic as Map<*, *>? ?: continue
            val graphicId = data["graphicId"] as String
            val graphicController = GraphicController(context, graphicId)
            graphicController.setSelectionPropertiesHandler(selectionPropertiesHandler)
            graphicController.interpretGraphicController(data, symbolVisibilityFilterController)
            graphicIdToController[graphicId] = graphicController
            graphicsAdded.add(graphicController.graphic)
        }

        graphicsOverlay.graphics.addAll(graphicsAdded)
    }

    fun changeGraphics(graphicsToChange: List<Any?>?) {
        if (graphicsToChange == null) {
            return
        }
        for (graphic in graphicsToChange) {
            val data = graphic as Map<*, *>? ?: continue
            val graphicId = data["graphicId"] as String
            val graphicController = graphicIdToController[graphicId] ?: continue
            graphicController.interpretGraphicController(data, symbolVisibilityFilterController)
        }
    }

    fun removeGraphics(graphicIdsToRemove: List<Any>?) {
        if (graphicIdsToRemove == null) {
            return
        }
        val graphicsToRemove = ArrayList<Graphic>(graphicIdsToRemove.size)
        for (rawGraphicId in graphicIdsToRemove) {
            val graphicId = rawGraphicId as String
            val graphicController = graphicIdToController.remove(graphicId)
            if (graphicController != null) {
                graphicsToRemove.add(graphicController.graphic)
            }
        }
        graphicsOverlay.graphics.removeAll(graphicsToRemove)
    }


}

File: ios/Classes/Mapping/Symbology/GraphicsController.swift
//
// Created by Valentin Grigorean on 29.05.2024.
//

import Foundation
import ArcGIS

class GraphicsController : NSObject, SymbolsController{
    private var graphicIdToController = Dictionary<String, GraphicMarkerController>()
    private let graphicsOverlay: GraphicsOverlay

    private let methodChannel: FlutterMethodChannel

    var selectionPropertiesHandler: SelectionPropertiesHandler?

    var symbolVisibilityFilterController: SymbolVisibilityFilterController?

    init(methodChannel: FlutterMethodChannel,
         graphicsOverlay: GraphicsOverlay
    ) {
        self.methodChannel = methodChannel
        self.graphicsOverlay = graphicsOverlay
    }

    func addGraphics(graphicsToAdd: [[String: Any]]) {
        for graphic in graphicsToAdd {
            let graphicId = graphic["graphicId"] as! String
            let controller = GraphicMarkerController(graphicId: graphicId)
            controller.selectionPropertiesHandler = selectionPropertiesHandler
            graphicIdToController[graphicId] = controller
            updateGraphic(data: graphic, controller: controller)
            controller.add(graphicsOverlay: graphicsOverlay)
                
        }
    }

    func changeGraphics(graphicsToChange: [[String: Any]]) {
        for graphic in graphicsToChange {
            let graphicId = graphic["graphicId"] as! String
            guard let controller = graphicIdToController[graphicId] else {
                continue
            }
            updateGraphic(data: graphic, controller: controller)
        }
    }

    func removeGraphics(graphicIdsToRemove: [String]) {
        for graphicId in graphicIdsToRemove {
            guard let controller = graphicIdToController[graphicId] else {
                continue
            }
            symbolVisibilityFilterController?.removeGraphicsController(graphicController: controller)
            controller.remove(graphicsOverlay: graphicsOverlay)
            graphicIdToController.removeValue(forKey: graphicId)
        }
    }

    

    private func updateGraphic(data: [String: Any], controller: GraphicMarkerController) {
        if let geometry = data["geometry"] as? [String: Any] {
            controller.geometry = Geometry.fromFlutter(data: geometry)
        }
        if let symbol = data["symbol"] as? [String:Any] {
            controller.symbol = SymbolFactory.createSymbol(data: symbol)
        }

        if let consumeTapEvents = data["consumeTapEvents"] as? Bool {
            controller.consumeTapEvents = consumeTapEvents
        }

        if let isSelected = data["isSelected"] as? Bool {
            controller.isSelected = isSelected
        }

        if let isVisible = data["isVisible"] as? Bool {
            controller.isVisible = isVisible
        }

        if let zIndex = data["zIndex"] as? Int {
            controller.zIndex = zIndex
        }
    }
}


extension GraphicsController: MapGraphicTouchDelegate {

    func canConsumeTaps() -> Bool {
        for (_, controller) in graphicIdToController {
            if controller.consumeTapEvents {
                return true
            }
        }
        return false
    }

    func didHandleGraphic(graphic: Graphic) -> Bool {
        guard let graphicId = graphic.attributes["graphicId"] as? String else {
            return false
        }
        if let currentMarker = graphicIdToController[graphicId] {
            guard currentMarker.consumeTapEvents else {
                return false
            }
        }
        methodChannel.invokeMethod("graphic#onTap", arguments: ["graphicId": graphicId])
        return true
    }
}
